File: 20171204-java-banking-kata/src/main/java/com/katas/Account.java
package com.katas;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class Account {

    private DateProvider dateProvider;
    private StatementPresenter statementPresenter;
    private List<Transaction> transactions;

    Account(DateProvider dateProvider, StatementPresenter statementPresenter) {
        this.dateProvider = dateProvider;
        this.statementPresenter = statementPresenter;
        this.transactions = new ArrayList<>();
    }

    public void deposit(int amount) {
        transactions.add(new Transaction(dateProvider.currentDate(), amount));
    }

    public String printStatement() {
        Integer balance = 0;

        List<StatementLine> statementLines = new ArrayList<>();
        for (Transaction transaction : transactions) {
            LocalDate date = transaction.getDate();
            Integer amount = transaction.getAmount();
            balance += amount;

            statementLines.add(new StatementLine(date, amount, balance));
        }

        return statementPresenter.printStatement(statementLines);
    }

}

File: 20171201-kotlin-banking-kata/src/main/kotlin/com/katas/Account.kt
package com.katas

import java.time.LocalDate
import java.time.format.DateTimeFormatter

class Account(private val dateProvider: DateProvider,
              private val statementPresenter: StatementPresenter) {

    private var transactions = mutableListOf<Transaction>()

    fun deposit(amount: Int) {
        val newBalance = balance() + amount
        transactions.add(Transaction(dateProvider.currentDate(), amount, newBalance))
    }

    fun withdraw(amount: Int) {
        val newBalance = balance() - amount
        transactions.add(Transaction(dateProvider.currentDate(), -amount, newBalance))
    }

    private fun balance(): Int {
        return if (transactions.isEmpty()) 0
        else transactions.last().balance
    }

    fun printStatement(): String = statementPresenter.print(transactions)
}

data class Transaction(val date: LocalDate, val amount: Int, val balance: Int)

interface StatementPresenter {
    fun print(transactions: List<Transaction>): String
}

class PrintStatementPresenter : StatementPresenter {
    private val header = "Date  Amount  Balance"
    private val formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy")

    override fun print(transactions: List<Transaction>): String {
        val rows = transactions.joinToString("", transform = this::printTransaction)

        return header + rows
    }

    private fun printTransaction(transaction: Transaction): String {
        val date = transaction.date.format(formatter)
        val amount = if (transaction.amount > 0) "+${transaction.amount}" else transaction.amount

        return "\n$date  $amount  ${transaction.balance}"
    }
}

